launcher/internal/gpu/driverinstaller.go

```diff
@@ -0,0 +1,161 @@
+func (di *DriverInstaller) InstallGPUDrivers(ctx context.Context) error {
+	gpuType, err := deviceinfo.GetGPUTypeInfo()
+	if err != nil {
+		return fmt.Errorf("failed to get the gpu type info: %v", err)
+	}
+
+	if !gpuType.OpenSupported() {
+		return fmt.Errorf("unsupported gpu type %s, please retry with one of the supported gpu types: %v", gpuType.String(), supportedGpuTypes)
+	}
+
+	ctx = namespaces.WithNamespace(ctx, namespaces.Default)
+	installerImageRef, err := getInstallerImageReference()
+	if err != nil {
+		di.logger.Printf("failed to get the installer container image reference: %v", err)
+		return err
+	}
+
+	di.logger.Printf("cos gpu installer version : %s", installerImageRef)
+	image, err := di.cdClient.Pull(ctx, installerImageRef, containerd.WithPullUnpack)
+	if err != nil {
+		return fmt.Errorf("failed to pull installer image: %v", err)
+	}
+
+	mounts := []specs.Mount{
+		{
+			Type:        "volume",
+			Source:      "/dev",
+			Destination: "/dev",
+			Options:     []string{"rbind", "rw"},
+		}, {
+			Type:        "volume",
+			Source:      "/",
+			Destination: "/root",
+			Options:     []string{"rbind", "rw"},
+		},
+	}
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		di.logger.Printf("cannot get hostname: %v", err)
+	}
+
+	container, err := di.cdClient.NewContainer(
+		ctx,
+		installerContainerID,
+		containerd.WithImage(image),
+		containerd.WithNewSnapshot(installerSnapshotID, image),
+		containerd.WithNewSpec(oci.WithImageConfig(image),
+			oci.WithPrivileged,
+			oci.WithProcessArgs("/cos-gpu-installer", "install", "-version=default", fmt.Sprintf("-host-dir=%s", InstallationHostDir)),
+			oci.WithAllDevicesAllowed,
+			oci.WithHostDevices,
+			oci.WithMounts(mounts),
+			oci.WithHostNamespace(specs.NetworkNamespace),
+			oci.WithHostNamespace(specs.PIDNamespace),
+			oci.WithHostHostsFile,
+			oci.WithHostResolvconf,
+			oci.WithEnv([]string{fmt.Sprintf("HOSTNAME=%s", hostname)})))
+	if err != nil {
+		return fmt.Errorf("failed to create gpu driver installer container: %v", err)
+	}
+
+	task, err := container.NewTask(ctx, cio.NewCreator(cio.WithStdio))
+	if err != nil {
+		return fmt.Errorf("failed to create gpu driver installation task: %v", err)
+	}
+
+	statusC, err := task.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to wait for gpu driver installation task: %v", err)
+	}
+
+	if err := task.Start(ctx); err != nil {
+		return fmt.Errorf("failed to start gpu driver installation task: %v", err)
+	}
+
+	status := <-statusC
+	code, _, _ := status.Result()
+	di.logger.Printf("Gpu driver installation task exited with status: %d\n", code)
+
+	err = remountAsExecutable(InstallationHostDir)
+	if err != nil {
+		return fmt.Errorf("failed to remount the installed drivers: %v", err)
+	}
+	return nil
+}
+
+func getInstallerImageReference() (string, error) {
+	installerImageRefBytes, err := exec.Command("cos-extensions", "list", "--", "--gpu-installer").Output()
```

Comment:
can we hardcode this ref? I run the command I saw "us.gcr.io/cos-cloud/cos-gpu-installer:v2.4.1"
I assume this won't change if in the same image?

Reply:
As we use image family (e.g. cos-113-lts`) field in cloudbuild, build will use the latest image from the family which may have the installer version updated and may break if we hardcode it.  So it is recommended to get the version using `cos-extensions` for given base cos image as `cos-gpu-installer` is not guaranteed to be backward compatible.

Reply:
This could be hardcoded by running this command at build time. That's because it will always be built on the same COS version. This should be fixed by GA; please add a TODO.